### PR TITLE
`HeteroLinear` with `segment_matmul`

### PR DESCRIPTION
**Pull request by puririshi98**

### CODE REVIEW

File: torch_geometric/nn/dense/linear.py [hunk 1]
```diff
@@ -161,42 +163,46 @@ def __repr__(self) -> str:
...
     to types
-
     .. math::
         \mathbf{x}^{\prime}_{\kappa} = \mathbf{x}_{\kappa}
         \mathbf{W}^{\top}_{\kappa} + \mathbf{b}_{\kappa}
-
     for type :math:`\kappa`.
     It supports lazy initialization and customizable weight and bias
     initialization.
-
```
Comment (rusty1s) [inline torch_geometric/nn/dense/linear.py:172]: Revert line break diff.

File: torch_geometric/nn/dense/linear.py [hunk 2]
```diff
@@ -161,42 +163,46 @@ def __repr__(self) -> str:
...
+        self._WITH_PYG_LIB = torch.cuda.is_available() and _WITH_PYG_LIB
+        self.is_sorted = is_sorted
         self.lins = torch.nn.ModuleList([
             Linear(in_channels, out_channels, **kwargs)
             for _ in range(num_types)
         ])
 
         self.reset_parameters()
+        if self._WITH_PYG_LIB:
```
Comment (rusty1s) [inline torch_geometric/nn/dense/linear.py:199]: Can't we instead initialize a `Parameter` with shape `[num_types, in_channels, out_channels]`?

File: torch_geometric/nn/dense/linear.py [hunk 3]
```diff
@@ -208,10 +214,21 @@ def forward(self, x: Tensor, type_vec: Tensor) -> Tensor:
...
-            out[mask] = lin(x[mask])
+        if self._WITH_PYG_LIB:
+            if not self.is_sorted:
+                if (type_vec[1:] < type_vec[:-1]).any():
+                    type_vec, perm = type_vec.sort()
+                    x = x[:, perm]
+            edge_type_ptr = torch.ops.torch_sparse.ind2ptr(
+                type_vec, self.num_types)
+            out = segment_matmul(x, edge_type_ptr, self.weights)
+            for i in range(self.num_types):
```
Comment (rusty1s) [inline torch_geometric/nn/dense/linear.py:225]: Any chance we can avoid the for-loop here? Something like `out += self.biases[type_vec]` should work as well.

**Review by rusty1s — APPROVED** [review]

File: torch_geometric/nn/dense/linear.py (hunk 2 above)
Comment (puririshi98) [reply torch_geometric/nn/dense/linear.py:199]: well i figured this way we preserve the kwargs and initiliazers of linear without having to re-code that specifically for the pyg_lib case

**Review by rusty1s — APPROVED** [review]

**Pull request merged**